contracts/SimpleSaleContract.sol

```diff
@@ -16,47 +23,140 @@ contract SimpleSaleContract is Ownable {
+        require(_tokensToBuy != 0, "No tokens bought");
+        _buyTokens(_tokensToBuy, _paymentToken);
+    }
 
-        uint256 tokensToPay = _tokensToBuy * paymentTokensExchangeRate[IERC20(_paymentToken)];
-        IERC20(_paymentToken).safeTransferFrom(msg.sender, address(this), tokensToPay);
+    /**
+     * @notice In exchange for _paymentToken, mintableTokens will be transferred to msg.sender at a specified exchange rate.
+     * Reverts if payment token is not specified.
+     * @param _tokensToBuy amount of mintable tokens to be bought
+     * @param _paymentToken the token that will be used for payment
+     */
+    function _buyTokens(uint256 _tokensToBuy, IERC20 _paymentToken) internal {
+        uint256 tokensToPay = _tokensToBuy 
+            * paymentTokensExchangeRate[_paymentToken].partsSell
+            / paymentTokensExchangeRate[_paymentToken].partsMint;
+        console.log(_paymentToken.balanceOf(msg.sender));
+        console.log(_paymentToken.allowance(msg.sender, address(this)));
+        _paymentToken.safeTransferFrom(msg.sender, address(this), tokensToPay);
         
         mintableToken.mint(msg.sender, _tokensToBuy);
 
-        emit SaleCompleted(msg.sender, _tokensToBuy, _paymentToken, tokensToPay);
+        emit TokenSaleCompleted(msg.sender, _tokensToBuy, _paymentToken, tokensToPay);
     }
 
     /**
-     * @notice admin function for setting payment token
+     * @notice admin function for setting payment token exchange rate
      * @param _paymentToken address of token to be set
-     * @param _exchangeRate exchange rate of the token in relation to mintable token
+     * @param _partsSellToken parts of the token that is sold to be converted into _partsMintToken
+     * @param _partsMintToken parts of the token that is minted in exchage of _partsSellToken
      */
-    function setPaymentTokenExchangeRate(address _paymentToken, uint256 _exchangeRate) external onlyOwner {
-        require(_paymentToken != address(0), "Payment token not set");
+    function setPaymentTokenExchangeRate(IERC20 _paymentToken, uint256 _partsSellToken, uint256 _partsMintToken) external onlyOwner {
+        require(address(_paymentToken) != address(0), "Payment token is invalid");
+        require(_partsSellToken != 0, "Invalid exchange rate");
+        require(_partsMintToken != 0, "Invalid exchange rate");
 
-        paymentTokensExchangeRate[IERC20(_paymentToken)] = _exchangeRate;
+        paymentTokensExchangeRate[_paymentToken].partsSell = _partsSellToken;
+        paymentTokensExchangeRate[_paymentToken].partsMint = _partsMintToken;
+
+        emit TokenExchangeRateSet(_paymentToken, _partsSellToken, _partsMintToken);
+    }
+
+    /**
+     * @notice admin function for unsetting payment token exchange rate
+     * @param _paymentToken address of token to be unset
+     */
+    function unsetPaymentTokenExchangeRate(IERC20 _paymentToken) external onlyOwner {
+        // Is it cheaper to call delete than x.a=0, x.b=0?
```

Comment:
it should be pretty much the same as delete sets the storage values to 0. Has to be checked.

Reply:
no, it's not cheaper! actually, 3 gas more expensive